experiments/process_sandbox/src/library_runner.cc

```diff
@@ -168,123 +170,90 @@ namespace
-  {
-    return GlobalPagemap::pagemap().get(p);
-  }
+  return done_bootstrapping;
+}
 
-  uint8_t ProxyPageMap::get(void* p)
+namespace sandbox
+{
+  snmalloc::CapPtr<void, snmalloc::CBChunk>
+  SnmallocGlobals::reserve(size_t size)
   {
-    return GlobalPagemap::pagemap().get(address_cast(p));
-  }
+    SNMALLOC_ASSERT(size >= sizeof(void*));
 
-  void ProxyPageMap::set_slab(snmalloc::Superslab* slab)
-  {
-    set(reinterpret_cast<uintptr_t>(slab), (size_t)CMSuperslab);
-  }
+    size = bits::align_up(size, sizeof(void*));
 
-  void ProxyPageMap::clear_slab(snmalloc::Superslab* slab)
-  {
-    set(reinterpret_cast<uintptr_t>(slab), (size_t)CMNotOurs);
-  }
+    size_t rsize = bits::next_pow2(size);
 
-  void ProxyPageMap::set_slab(snmalloc::Mediumslab* slab)
-  {
-    set(reinterpret_cast<uintptr_t>(slab), (size_t)CMMediumslab);
-  }
+    snmalloc::CapPtr<void, snmalloc::CBChunk> res{
+      reinterpret_cast<void*>(requestHostService(
+        MemoryProviderReserve, static_cast<uintptr_t>(rsize)))};
 
-  void ProxyPageMap::clear_slab(snmalloc::Mediumslab* slab)
-  {
-    set(reinterpret_cast<uintptr_t>(slab), (size_t)CMNotOurs);
+    return res;
   }
 
-  void ProxyPageMap::set_large_size(void* p, size_t size)
+  void SnmallocGlobals::Pagemap::set_metaentry(
+    LocalState*, snmalloc::address_t p, size_t size, snmalloc::MetaEntry t)
   {
-    size_t size_bits = bits::next_pow2_bits(size);
-    SANDBOX_DEBUG_INVARIANT(
-      is_inside_shared_memory(p, size),
-      "Setting large size for pointer {} in pagemap that is outside of the "
-      "sandbox range {}--{}",
-      p,
-      shared_memory_start,
-      shared_memory_end);
     requestHostService(
-      ChunkMapSetRange,
-      reinterpret_cast<uintptr_t>(p),
-      static_cast<uintptr_t>(size_bits));
+      MetadataSet,
+      static_cast<uintptr_t>(p),
+      static_cast<uintptr_t>(size),
+      reinterpret_cast<uintptr_t>(t.get_metaslab()),
+      t.get_remote_and_sizeclass());
+  }
+
+  template<>
+  snmalloc::CapPtr<void, snmalloc::CBChunk>
+  SnmallocGlobals::alloc_meta_data<snmalloc::Metaslab>(LocalState*, size_t size)
+  {
+    size = snmalloc::bits::next_pow2(size);
+    return private_asm.reserve<true, SnmallocGlobals::PrivatePagemap>(
+      nullptr, size);
+  }
+
+  template<>
+  snmalloc::CapPtr<void, snmalloc::CBChunk> SnmallocGlobals::alloc_meta_data<
+    snmalloc::CoreAllocator<sandbox::SnmallocGlobals>>(LocalState*, size_t size)
+  {
+    size = snmalloc::bits::next_pow2(size);
+    auto* result =
+      reinterpret_cast<void*>(requestHostService(MemoryProviderReserve, size));
+    return snmalloc::CapPtr<void, snmalloc::CBChunk>{result};
+  }
+
+  std::pair<snmalloc::CapPtr<void, snmalloc::CBChunk>, snmalloc::Metaslab*>
+  SnmallocGlobals::alloc_chunk(
+    SnmallocGlobals::LocalState*,
+    size_t size,
+    snmalloc::RemoteAllocator* remote,
+    snmalloc::sizeclass_t sizeclass)
+  {
+    // TODO: Check this doesn't actually write to the pagemap.
```

Comment:
It looks like it's OK if it does, now that it's pointing at the private pagemap?